desc/objectives/linear_objectives.py

```diff
@@ -408,6 +397,329 @@ def compute(self, params, constants=None):
         return jnp.dot(self._A, params["Z_lmn"]) - params["Zb_lmn"]
 
 
+class SectionRSelfConsistency(_Objective):
+    """Ensure that the boundary and interior surfaces are self-consistent.
+
+    Note: this constraint is automatically applied when needed, and does not need to be
+    included by the user.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    surface_label : float, optional
+        Surface to enforce boundary conditions on. Defaults to Equilibrium.surface.rho
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _linear = True
+    _fixed = False
+    _units = "(m)"
+    _print_value_fmt = "R cross-section self consistency error: {:10.3e} "
+
+    def __init__(
+        self,
+        eq,
+        name="self_consistency section R",
+    ):
+        super().__init__(
+            things=eq,
+            target=0,
+            bounds=None,
+            weight=1,
+            normalize=False,
+            normalize_target=False,
+            name=name,
+        )
+
+    def build(self, use_jit=False, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        modes = eq.xsection.R_basis.modes
+        self._dim_f = eq.xsection.R_basis.num_modes
+        self._A = np.zeros((self._dim_f, eq.R_basis.num_modes))
+
+        if eq.xsection.zeta == 0:
+            for i, (l, m, n) in enumerate(modes):
+                j = np.argwhere(
+                    np.logical_and(
+                        (eq.R_basis.modes[:, :-1] == [l, m]).all(axis=1),
+                        eq.R_basis.modes[:, 2] >= 0,
+                    )
+                )
+                self._A[i, j] = 1
+        elif eq.xsection.zeta == np.pi:
```

Comment:
does it have to be literally pi? or pi/NFP?

ie, if NFP > 1 then zeta=pi might be in a different field period and might mean something different than the "half period" cross section

Reply:
It is actually pi/NFP but I made the functions take in zeta as just angle and then divide by eq.NFP inside, so the calculations are correct. For the version to be merged, I can change it back to proper zeta/eq.NFP .

Reply:
I tried that once but since zeta=0 and pi (pi/nfp actually) has proper lambda gauge thing, it was being weird. So I just chose 0 and pi as possible values for poincare.